traversal/planner/GraphPlanner.java

```diff
@@ -291,11 +294,31 @@ private void optimise(long timeLimitMillis) {
         isOptimising.set(false);
     }
 
+    private void linearise() {
+        Set<PlannerVertex<?>> visited = new HashSet<>();
+        List<PlannerVertex<?>> stack = vertices.values().stream().filter(PlannerVertex::isStartingVertex).collect(Collectors.toList());
```

Comment:
in this case, you should actually be making sure that you're using a `LinkedList` because of the operations you'r doing on it.

Also notice that `visited` and `stack` are not symmetrical... one is an abstract term and one is a contextualised term. Can you find a contextualised term for `stack`?